### PR TITLE
refactor: Replace "cache-control" extension with plugin.

### PR DESCRIPTION
**Pull request by abernix**
Similar to 6009d8a00278e (#3991), which migrated the tracing package to the plugin API, this commit introduces a `plugin` (named) export from the `apollo-cache-control` module.  It similarly accomplishes that by duplicating the behavior of the `CacheControlExtension` which leveraged the soon-to-be-deprecated `graphql-extensions`.  The functionality, again, is intended to be identical in spirit.

As with the afformentioned commit, since the delta of the commits was otherwise even _more_ confusing, I've left the `CacheControlExtension` present and exported and will remove it in a subsequent commit.

Briefly summarizing what the necessary changes were:

1. We no longer use a class instance to house the extension, which was necessitated by the `graphql-extensions` API.  This means that uses of `this` have been replaced with function scoped variables by the same name.
2. The logic which actually does the formatting (previously handled by the `format` method in `graphql-extension`, now takes place within the plugin API's `willSendResponse` method.
3. Rather than leaning on plugin-specific behavior from within the request pipeline, the `apollo-cache-control` plugin now makes a number of assertions throughout its various life-cycle methods to ensure that the `overallCachePolicy` is calculated.
4. Switch tests to use the new `pluginTestHarness` method for testing plugins which was introduced in eec87a6c6dead7fd (#3990).

### CODE REVIEW

File: packages/apollo-server-core/src/ApolloServer.ts [hunk 1]
```diff
@@ -790,7 +770,37 @@ export class ApolloServerBase {
     // at the end of that list so they take precidence.
     // A follow-up commit will actually introduce this.
 
+    // Enable cache control unless it was explicitly disabled.
+    if (this.config.cacheControl !== false) {
```
Comment (trevor-scheer) [inline packages/apollo-server-core/src/ApolloServer.ts:774]: With `cacheControl` implemented as a plugin now, do you anticipate we'll remove this configuration point in a future major (v4)?

File: packages/apollo-server-core/src/requestPipeline.ts [hunk 2]
```diff
@@ -390,20 +384,6 @@ export async function processGraphQLRequest<TContext>(
       }
     }
 
-    if (cacheControlExtension) {
```
Comment (trevor-scheer) [inline packages/apollo-server-core/src/requestPipeline.ts:393]: Love to see it!

File: packages/apollo-server-core/src/ApolloServer.ts (hunk 1 above)
Comment (abernix) [reply packages/apollo-server-core/src/ApolloServer.ts:751]: Possibly.  I feel like our `cacheControl` integration might get tighter over time, but it's certainly possible we could have configuration be entirely composable via plugins.  I don't imagine we'd sacrifice this particular convenience plugin initialization anytime soon though.

File: packages/apollo-cache-control/src/index.ts [hunk 3]
```diff
@@ -49,164 +47,148 @@ declare module 'apollo-server-types' {
...
+          info.cacheControl = {
+            setCacheHint: (hint: CacheHint) => {
+              addHint(hints, info.path, hint);
+            },
+            cacheHint: hint,
+          };
+        },
+      }),
+
+      responseForOperation() {
```
Comment (glasser) [inline packages/apollo-cache-control/src/index.ts:127]: @abernix Do you have any memory of why this is here? This runs before execution, so I don't see how the policy it calculates could ever be non-trivial. I don't see anything that obviously matches it on the "before" side of this PR. Tests pass without it (on my PR that refactors some stuff).

**Pull request merged**